### ISSUE TITLE
Index based DataSet creation

### ISSUE THREAD

**Issue by Tradunsky** [nanne-aben/strictly_typed_pandas#8]
In Pandas it is common to create DataFrame as follows:
```python
from pandas import DataFrame
DataFrame(data=[("Nanne", "Aben"), ("Slava", "Tradunsky")], columns=["First name", "Last name"])
```
```
  First name  Last name
0      Nanne       Aben
1      Slava  Tradunsky
```
However, when trying to do an analogy in DataSet:
```python
from dataclasses import dataclass
@dataclass
class Person:
    first_name: str
    last_name: str


DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")])
```
```
...
File "python3.9/site-packages/strictly_typed_pandas/dataset.py", line 119, in _continue_initialization
validate_schema(schema_expected, schema_observed)
File "python3.9/site-packages/strictly_typed_pandas/validate_schema.py", line 30, in validate_schema
_check_names(set(schema_expected.keys()), set(schema_observed.keys()))
File "python3.9/site-packages/strictly_typed_pandas/validate_schema.py", line 37, in _check_names
raise TypeError(
TypeError: Data contains the following columns not present in schema: {0, 1}
```
Assuming schema is known due to generic class Person, I am not passing it as a dictionary nor giving `columns`.

Would be great to have the least resistance path to migrate existing code from DataFrame to DataSet without changing `data` parameter shape too much. 🙏🏻

**Comment by nanne-aben** [comment]
Hi Slava!

Thanks for your feedback! I think the most straightforward solution would be to define the DataSet as:

```{python}
from strictly_typed_pandas import DataSet

class Person:
    first_name: str
    last_name: str

DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")], columns=["first_name", "last_name"])
```

You're right that in this case you should be able to determine the column names from the Person class. You could do something like this (though it doesn't look great, I admit):
```{python}
from strictly_typed_pandas import DataSet

class Person:
    first_name: str
    last_name: str

DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")], columns=Person.__annotations__.keys())
```

While I agree that it would be neat if strictly_typed_pandas would take care of this for you, there is one problem with that: we don't actually assume anywhere that the ordering of variables in the schema is the same as the ordering in the columns. For example, this would also work:

```{python}
from strictly_typed_pandas import DataSet

# switching first name and last name
class Person:
    last_name: str
    first_name: str

DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")], columns=["first_name", "last_name"])
```

I don't really want to change that behaviour. One of the advantages of order independence is that we could use the Prototype class down the line to move a lot more type checking into the linting part (using the Prototype type), rather than relying on run-time checks.

**Comment by Tradunsky** [comment]
Hi @nanne-aben,

Thank you for your answer and suggestions!

That's indeed an issue with sorting I did not realize originally! Great that you pointed it out. 

Seems like the suggestion you made, would explicitly define the order, but indeed the class design looks not ideal:
```
DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")], columns=Person.__annotations__.keys())
```
____

I started to think about linting more and would like to hear your opinion:
Whenever I create a `pandas.DataFrame` it has clear field type variations that it can be constructed from.
<img width="1258" alt="image" src="https://user-images.githubusercontent.com/4522842/184522817-f616c602-e776-4a34-806e-3c1bbcd891b1.png">
This makes class design complete, as I can construct DataFrame of the given data shape and have linting time type checking:
<img width="1260" alt="image" src="https://user-images.githubusercontent.com/4522842/184523126-88718d5b-087b-4e8c-aabc-d0ab3b111512.png">
```
       0          1
0  Nanne       Aben
1  Slava  Tradunsky
```

Whereas `DataSet` class design does not give any clue about what parameter types it expects (show parameters command does nothing). I believe this is primarily because `DataSetBase` eliminates all work done in pandas to introduce type hints for better linting to find errors earlier than runtime:
```python
class DataSetBase(pd.DataFrame, ABC):
    def __init__(self, *args, **kwargs) -> None: # here all DataFrame type hints are effectively shallowed
``` 

But since the main purpose of the library is to have generic specialized content of the collection data structure (analogy to List[Person], Set[Person]), `DataSet[Person]` design is intuitively expected to represent its content, however, when a user tries to follow the common class design and construct an object, it does not show linting type error, but throws a bit confusing runtime error:
```python
DataSet[Person](data=[("Nanne", "Aben"), ("Slava", "Tradunsky")])
```
```
TypeError: Data contains the following columns not present in schema: {0, 1}
```

WDYT?

**Comment by nanne-aben** [comment]
Yeah, I've noticed that the parameter type hints (and actually, all documentation about parameters) goes out of the window once you subclass something. I wish I could inherit those somehow, but I have no idea how to do this. If you have suggestions, those would be very welcome!

With respect to the error: since you didn't define column names, pandas automatically creates a DataFrame with columns ['0', '1']. Strictly typed pandas then detects that those are different from what you've defined in the schema and gives the error you see. Maybe the error could be a bit more clear indeed. We could check in DataSetBase whether the input is a list of tuples, and whether a columns argument has been provided, and throw a warning accordingly. What do you think? Would you like to make a PR?
